File: dtbi-2023/21_sauteMouton/algo_resolution/sm.cpp

#include <iostream>
#include <queue>
#include <set>
#include <string>
#include <vector>

using namespace std;

bool IsValidMove(const string& current, const string& next) {
    int diff_count = 0, last_diff = -1;
    for (int i = 0; i < current.size(); ++i) {
        if (current[i] != next[i]) {
            diff_count++;
            if (last_diff == -1) last_diff = i;
            else if (abs(i - last_diff) > 2) return false;
        }
    }
    return diff_count == 2;
}

vector<string> GetNextStates(const string& state) {
    vector<string> next_states;
    int empty_index = state.find('V');

    for (int i : {-1, 1, -2, 2}) {
        if (empty_index + i >= 0 && empty_index + i < state.size()) {
            string next_state = state;
            swap(next_state[empty_index], next_state[empty_index + i]);
            if (IsValidMove(state, next_state)) {
                next_states.push_back(next_state);
            }
        }
    }

    return next_states;
}

vector<string> SolveSheepPuzzle() {
    string initial_state = "BBBVNNN";
    string target_state = "NNNVBBB";
    queue<pair<string, vector<string>>> q;
    set<string> visited;

    q.push({initial_state, {initial_state}});
    visited.insert(initial_state);

    while (!q.empty()) {
        auto [current_state, path] = q.front();
        q.pop();

        if (current_state == target_state) {
            return path;
        }

        for (const auto& next_state : GetNextStates(current_state)) {
            if (!visited.count(next_state)) {
                visited.insert(next_state);
                vector<string> new_path = path;
                new_path.push_back(next_state);
                q.push({next_state, new_path});
            }
        }
    }

    return {};
}

int main() {
    vector<string> solution = SolveSheepPuzzle();
    for (const string& state : solution) {
        cout << state << endl;
    }
    return 0;
}

File: dtbi-2023/21_sauteMouton/algo_resolution/sm.py
from collections import deque

def is_valid_move(state, pos, move):
    """ Check if a move is valid. """
    if move == -1:
        # Move left
        return pos > 0 and state[pos-1] != state[pos]
    elif move == 1:
        # Move right
        return pos < len(state) - 1 and state[pos+1] != state[pos]
    elif move == -2:
        # Jump left
        return pos > 1 and state[pos-2] != state[pos] and state[pos-1] != state[pos]
    elif move == 2:
        # Jump right
        return pos < len(state) - 2 and state[pos+2] != state[pos] and state[pos+1] != state[pos]
    return False

def get_next_states(state):
    """ Get all possible next states from the current state. """
    next_states = []
    empty_pos = state.find('0')
    moves = [-1, 1, -2, 2]  # Left, Right, Jump left, Jump right

    for move in moves:
        if is_valid_move(state, empty_pos, move):
            new_pos = empty_pos + move
            new_state = list(state)
            new_state[empty_pos], new_state[new_pos] = new_state[new_pos], new_state[empty_pos]
            next_states.append("".join(new_state))

    return next_states

def find_solution(start_state, end_state):
    """ Find the solution using Breadth-First Search. """
    queue = deque([(start_state, [start_state])])  # State and path
    visited = set()

    while queue:
        current_state, path = queue.popleft()
        if current_state == end_state:
            return path  # Found the solution

        for next_state in get_next_states(current_state):
            if next_state not in visited:
                visited.add(next_state)
                queue.append((next_state, path + [next_state]))

    return None  # No solution found

# Starting state and ending state
start_state = "BBB0NNN"
end_state = "NNN0BBB"

# Find the solution
solution = find_solution(start_state, end_state)

print(solution)